web/src/copy_and_paste.ts

```diff
@@ -472,6 +489,122 @@ export function paste_handler_converter(paste_html: string): string {
+                node.classList.contains("katex-display") &&
+                !node.closest(".katex-display > .katex-display")
+            ) {
+                return true;
+            }
+
+            return false;
+        },
+        replacement(content: string, node: Node) {
+            assert(node instanceof HTMLElement);
+            let math_block_markdown = "```math\n";
+            if (node !== null) {
+                const parent = node.parentElement!;
+                if (processed_math_block_parents.has(parent)) {
+                    return "";
+                }
+                processed_math_block_parents.add(parent);
+                let consecutive_empty_display_count = 0;
+                for (const child of parent.children) {
+                    const annotation_element = child.querySelector(
+                        '.katex-mathml annotation[encoding="application/x-tex"]',
+                    );
+
+                    if (annotation_element?.textContent) {
+                        const katex_source = annotation_element.textContent.trim();
+                        math_block_markdown += katex_source + "\n\n";
+                        consecutive_empty_display_count = 0;
+                    } else {
+                        // Handling cases where math block is selected directly without any preceding text.
+                        // The initial katex display doesn't have the annotation when selection is done in this manner.
+                        if (
+                            child.classList.contains("katex-display") &&
+                            child.querySelector("math")?.textContent !== ""
+                        ) {
+                            math_block_markdown += content;
+                            continue;
+                        }
+                        if (consecutive_empty_display_count === 0) {
+                            math_block_markdown += "\n\n\n";
+                        } else {
+                            math_block_markdown += "\n\n";
+                        }
+                        consecutive_empty_display_count += 1;
+                    }
+                }
+            }
+            math_block_markdown = math_block_markdown.slice(0, -1); // Don't add extra newline add the end
+            return (math_block_markdown += "```");
+        },
+    });
+
+    turndownService.addRule("katex-inline-math", {
+        filter(node) {
+            // Allow the intermediate text blocks, so that they don't get appended to the end.
+            if (node.classList.contains("intermediate-katex-text-node")) {
+                return true;
+            }
+            if (node.classList.contains("katex") && !node.classList.contains("katex-display")) {
+                // Should explicitly be an inline expression
+                const parent = node.parentElement;
+                if (parent?.classList.contains("katex-display")) {
+                    return false;
+                }
```

Comment:
Are there elements with a `katex` class inside a `katex-display`? I feel like maybe a block comment explaining the nested structure of these somewhere in this file would be helpful.

Reply:
Well, I'd make it brief, but it's a fairly complex nested structure and if written down concisely here, it'll very much make this code more readable for someone who is trying to verify the logic.

Something at the level of detail of this (field names made up, I'm not looking at the HTML structure right now)
```
// This is the structure of a display math block. 
div.katex-display
   div.katex-first-thing
        annotation.annotation-class-name
   ... (other fields we don't care about)
   div.katex-other-thing
```